doc/rst/tools/chapel-py/chapel-py.rst

```diff
@@ -0,0 +1,475 @@
+  certain utility functions. For instance, it provides the ``preorder`` and
+  ``postorder``, traversal iterators, and a ``match`` function to perform AST
+  pattern matching.
+* The ``chapel.replace`` module provides the "replacer API", which is intended
+  to perform transformations on existing Chapel files for various reasons. For
+  instance, deprecations and syntax changes can be performed automatically using
+  the replacer API, by finding AST patterns and performing string substitution.
+* The ``chapel.lsp`` module provides a few helpers to transform Dyno types to LSP
+  (Language Server Protocol) types.
+* The ``chapel.core`` module which provides the parts of this library that are
+  bridged from C++. Most users should not need to use this library, as all symbols are forwarded to ``chapel``.
+
+``chapel``
+~~~~~~~~~~
+
+The main entry point to the Chapel Python API is the ``Context`` object. This
+is a wrapper around the C++ construct of the same name. The Context in 'dyno'
+is responsible of memoizing computations, interning strings, and more. The
+Python wrapper around this object has methods that can be used to get started
+with processing Chapel programs; one such method is ``parse``, which accepts
+a file path and returns a list of Chapel AST nodes that represent this file.
+
+At compile-time, ``chapel-py`` generates a class hierarchy of Python objects,
+with base class ``AstNode``. A class is created for each of the AST nodes understood
+by Dyno. Thus, one might write:
+
+.. code-block:: python
+
+   from chapel import *
+
+   ctx = Context()
+   asts = ctx.parse("myfile.chpl")
+   print("it's a module" if isinstance(asts[0], Module) else "it's not a module")
+
+Here, the Python ``isinstance`` builtin is used to determine if the first
+AST node returned by ``parse`` is a module (it should be!)
+
+Because the AST nodes form a class hierarchy, it's possible to perform a more
+general check; the following example prints "It's a loop!" for all loops
+(``do``- ``while``, ``for``, ``foreach``, etc.), and "It's not a loop..." otherwise.
+
+.. code-block:: python
+
+   def check(node):
+       if isinstance(node, Loop):
+           print("It's a loop!")
+       else:
+           print("It's not a loop...")
+
+The AST nodes from the Python API are iterable; iterating over a node yields
+its children. Thus, one might print all the declarations in a module as follows:
+
+.. code-block:: python
+
+   def print_decls(mod):
+       for child in mod:
+           if isinstance(child, NamedDecl):
+               print(child.name())
+
+Full AST traversals can be built on top of this functionality. Below
+is the definition for ``postorder`` from the ``chapel`` module:
+
+.. code-block:: python
+
+   def postorder(node):
+       """
+       Recursively visit the given AST node, going in post-order (children-then-parent)
+       """
+       for child in node:
+           yield from preorder(child)
```

Comment:
The `postorder` in `tools/chapel-py/src/chapel/__init__.py` yields from `postorder`, not from `preorder`.

Reply:
This is funny. We fixed this bug in the library, but the copy-paste into the README.md (now transferred into the RST) remained unfixed.

Reply:
Fixed in the current draft